### PR TITLE
feat(Chat): Added chat censorship spoofing & duplicate message detection

### PR DESCRIPTION
**Pull request by ghost**
Was messing around with chat things and decided that I enjoy having these features in whatever menu I'm using. I'm almost positive I included everything correctly (Based on Yim standards). Let me know if there's anything I should fix or add, and I will get to it when I can. There's probably a better way of doing this, but this what I was able to come up with in a short amount of time.

### CODE REVIEW

File: BigBaseV2/src/core/globals.hpp [hunk 1]
```diff
@@ -67,6 +67,8 @@ namespace big
 			pair net_array_error{};
 			pair network_player_mgr_shutdown{};
 
+			pair chat_recieve{};
```
Comment (Yimura) [inline BigBaseV2/src/core/globals.hpp]: recieve is incorrect, it should be receive

File: BigBaseV2/src/core/globals.hpp [hunk 2]
```diff
@@ -291,6 +296,9 @@ namespace big
 			g->notifications.gta_thread_start.log = j["notifications"]["gta_thread_start"]["log"];
 			g->notifications.gta_thread_start.notify = j["notifications"]["gta_thread_start"]["notify"];
 
+			g->notifications.chat_recieve.log = j["notifications"]["chat_recieve"]["log"];
+			g->notifications.chat_recieve.notify = j["notifications"]["chat_recieve"]["notify"];
+
```
Comment (Yimura) [inline BigBaseV2/src/core/globals.hpp]: recieve is incorrect, it should be receive

File: BigBaseV2/src/core/globals.hpp [hunk 3]
```diff
@@ -496,8 +505,9 @@ namespace big
 					"notifications", {
 						{ "gta_thread_kill", return_notify_pair(g->notifications.gta_thread_kill) },
 						{ "gta_thread_start", return_notify_pair(g->notifications.gta_thread_start) },
-						{"net_array_error", return_notify_pair(g->notifications.net_array_error)},
+						{ "net_array_error", return_notify_pair(g->notifications.net_array_error) },
 						{ "network_player_mgr_shutdown", return_notify_pair(g->notifications.network_player_mgr_shutdown) },
+						{ "chat_recieve", return_notify_pair(g->notifications.chat_recieve) },
```
Comment (Yimura) [inline BigBaseV2/src/core/globals.hpp]: recieve is incorrect, it should be receive

File: BigBaseV2/src/hooking.cpp [hunk 4]
```diff
@@ -59,7 +59,12 @@ namespace big
 		// Player Has Joined
 		m_player_has_joined_hook("PHJ", g_pointers->m_player_has_joined, &hooks::player_join),
 		// Player Has Left
-		m_player_has_left_hook("PHL", g_pointers->m_player_has_left, &hooks::player_leave)
+		m_player_has_left_hook("PHL", g_pointers->m_player_has_left, &hooks::player_leave),
+
+		// Chat recieve
+		m_chat_receive_hook("CR", g_pointers->m_chat_receive, &hooks::chat_receive),
```
Comment (Yimura) [inline BigBaseV2/src/hooking.cpp]: recieve is incorrect, it should be receive

File: BigBaseV2/src/hooking.cpp [hunk 5]
```diff
@@ -105,13 +110,21 @@ namespace big
...
 		m_enabled = true;
 	}
 
 	void hooking::disable()
 	{
 		m_enabled = false;
 
+		m_censor_chat_text_hook.disable();
+
+		m_chat_receive_hook.disable();
```
Comment (Yimura) [inline BigBaseV2/src/hooking.cpp:126]: recieve is incorrect, it should be receive

File: BigBaseV2/src/hooking.cpp [hunk 6]
```diff
@@ -105,13 +110,21 @@ namespace big
 
 		m_send_net_info_to_lobby.enable();
 
+		m_chat_receive_hook.enable();
```
Comment (Yimura) [inline BigBaseV2/src/hooking.cpp:113]: recieve is incorrect, it should be receive

File: BigBaseV2/src/hooks/chat_recieve.cpp [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+	{
+		if (msg == previous_message)
+		{
+			std::string message = fmt::format("Duplicate message purged: {}", msg);
+
+			if (g->notifications.chat_recieve.log)
+				LOG(WARNING) << message;
+
+			if (g->notifications.chat_recieve.notify)
+				g_notification_service->push_warning("Protections", message);
```
Comment (Yimura) [inline BigBaseV2/src/hooks/chat_recieve.cpp]: Be more descriptive when logging / notifying the user.

**Review by Yimura — CHANGES_REQUESTED** [review]

**Comment by cl52902** [comment]
Lol I'll fix my typo and make the notification more descriptive. You should fix the dozens of other typos in your variables too though 😉.

**Comment by Yimura** [comment]
> Lol I'll fix my typo and make the notification more descriptive. You should fix the dozens of other typos in your variables too though 😉.

You are right my English isn't perfect, though I think it most of the text in YimMenu is correct. Make an issue if you see any typo's you'd like changed.

File: BigBaseV2/src/hooking.cpp (hunk 4 above)
Comment (ghost) [reply BigBaseV2/src/hooking.cpp]: It already is

File: BigBaseV2/src/hooking.cpp (hunk 5 above)
Comment (ghost) [reply BigBaseV2/src/hooking.cpp:126]: It already is

File: BigBaseV2/src/hooking.cpp (hunk 6 above)
Comment (ghost) [reply BigBaseV2/src/hooking.cpp:113]: It already is

**Comment by ghost** [comment]
Fixed typos. I'm not sure what more you want from the notification / logging. You could have it display who sent the message, but I didn't look into that.

**Review by Yimura — APPROVED** [review]

**Comment by Yimura** [comment]
Will check it when I'm back on my Windows work PC.

Edit:
Lmao just saw this code contains desync kick :blush:, don't worry though, I'll edit it out in the final merge. @LiamD-Flop has added desync kick several times by accident to YimMenu.

**Comment by GibsonHF** [comment]
Any way to add commands for other players to this for example /spawn [vehicle] that way others can benefit from our mods, maybe /gift too so they can gift themselves the vehicle to keep it

**Comment by ghost** [comment]
> Any way to add commands for other players to this for example /spawn [vehicle] that way others can benefit from our mods, maybe /gift too so they can gift themselves the vehicle to keep it

I know this is an old comment and a closed commit, but you would need to find a way to get the players info from this hook. If you do that, it's as simple as checking for keywords in the message and spawning whatever you want at the players coordinates based on the command they sent.

**Pull request closed without merging**